src/systems.rs

```diff
@@ -151,27 +153,6 @@ pub fn download_slippy_tiles(
 
         for x in min_x..=max_x {
             for y in min_y..=max_y {
-                // Check concurrent download limit
-                if active_downloads.0.load(Ordering::Relaxed)
-                    >= slippy_tiles_settings.max_concurrent_downloads
-                {
-                    warn!("Max concurrent downloads reached, buffering tile download");
-                    rate_limiter.buffer_request(
```

Comment:
[MEMO] We don't check whether the tiles are cached at this point, so this `buffer_request` call forces all tiles to be downloaded from the internet.

Even worse, `AtomicUsize` is difficult to use correctly. For example, we must specify the appropriate `Ordering` instead of always using `Ordering::SeqCst`. So I decided to use `Semaphore` instead of `AtomicUsize` (58da9dc9cde3bb251969df3ee7f50047a9f74f14) .